### PR TITLE
[DWARF] Add option to add linkage_names to call_origin declaration refs

### PR DESCRIPTION
**Pull request by OCHyams**
If `-mllvm -add-linkage-names-to-external-call-origins` is true then add
`DW_AT_linkage_name` attributes to `DW_TAG_subprogram` DIEs referenced by
`DW_AT_call_origin` attributes that would otherwise be omitted.

---

A debugger may use `DW_TAG_call_origin` attributes to determine whether any
frames in a call stack are missing due to optimisations (e.g. tail calls).

For example, say `a()` calls `b()` tail-calls `c()`, and you stop in your debugger
in `c()`:

The callstack looks like this:

    #0 c() <- stopped in this frame
    #1 a()

Looking "up" from `c()`, call site information can be found in `a()`. This includes
a DW_AT_call_origin referencing `b()`'s subprogram DIE, which means the call at
this call site was to `b()`, not `c()` where we are currently stopped. This
indicates `b()`'s frame has been lost due to optimisation (or is misleading due
to ICF).

This patch makes it easier for a debugger to check whether the referenced
DIE describes the target function or not, for example by comparing the referenced
function name to the current frame.

---

There's already an option to apply DW_AT_linkage_name in a targeted manner: `-dwarf-linkage-names=Abstract`, which limits adding DW_AT_linkage_names to abstract subprogram DIEs (this is default for SCE tuning).

The new flag shouldn't affect non-SCE-tuned behaviour whether it is enabled or not because the non-SCE-tuned behaviour is to always add linkage names to subprogram DIEs, AFAICT.

Enabling this feature alongside `-dwarf-linkage-names=Abstract` has minimal impact on size of CTMark RelWithDebInfo binaries. There's a geomean increase in size by < 0.1%.

### CODE REVIEW

File: llvm/lib/CodeGen/AsmPrinter/DwarfCompileUnit.cpp [hunk 1]
```diff
@@ -35,13 +35,28 @@
 #include "llvm/Target/TargetLoweringObjectFile.h"
 #include "llvm/Target/TargetMachine.h"
 #include "llvm/Target/TargetOptions.h"
+#include "llvm/Support/CommandLine.h"
```
Comment (pogo59) [inline llvm/lib/CodeGen/AsmPrinter/DwarfCompileUnit.cpp:38]: clang-format reminds me this is not in correct order.

File: llvm/test/DebugInfo/X86/call-origin-linkage-names.ll [hunk 2]
```diff
@@ -0,0 +1,96 @@
+; RUN: llc %s --filetype=obj -o - -dwarf-linkage-names=Abstract -add-linkage-names-to-declaration-call-origins=false \
+; RUN: | llvm-dwarfdump - | FileCheck %s --check-prefixes=COMMON,DISABLE --implicit-check-not=DW_AT_linkage_name
+; RUN: llc %s --filetype=obj -o - -dwarf-linkage-names=Abstract -add-linkage-names-to-declaration-call-origins=true \
+; RUN: | llvm-dwarfdump - | FileCheck %s --check-prefixes=COMMON,ENABLE --implicit-check-not=DW_AT_linkage_name
+
```
Comment (pogo59) [inline llvm/test/DebugInfo/X86/call-origin-linkage-names.ll:5]: Do we want more RUN lines showing default varies with tuning? I'm inclined to think it's not super necessary but curious what other reviewers think.

**Review by pogo59 — COMMENTED** [review]
LGTM for our debugger's needs, but I'll hold off approving to give others a chance to look at it.

File: llvm/lib/CodeGen/AsmPrinter/DwarfCompileUnit.cpp [hunk 3]
```diff
@@ -35,13 +35,28 @@
...
 
 using namespace llvm;
 
+/// Query value using AddLinkageNamesToDeclCallOriginsForTuning.
+cl::opt<cl::boolOrDefault> AddLinkageNamesToDeclCallOrigins(
+    "add-linkage-names-to-declaration-call-origins", cl::Hidden,
+    cl::desc("Add DW_AT_linkage_name to function declaration DIEs "
+             "referenced by DW_AT_call_origin attributes. Enabled by default "
+             "for -gsce debugger tuning."));
+
```
Comment (dwblaikie) [inline llvm/lib/CodeGen/AsmPrinter/DwarfCompileUnit.cpp:52]: I think we /generally/ try not to use/add raw cl::opts, and instead pass things through the driver/frontend flag handling? (I might be wrong, if there's substantial precedent, let me know)

& not sure how this sits with @pogo59's general premise that debugger tuning shouldn't be reachable only through the tuning flags, but should be independently accessible (does independently accessible mean supported in the driver flags, or is -Xclang/cc1/mllvm flags sufficient to meet that requirement?)

**Comment by dwblaikie** [comment]
> The new flag shouldn't change anything upstream whether it is enabled or not because the non-SCE-tuned behaviour upstream is to always add linkage names to subprogram DIEs, AFAICT.

By "upstream" you mean "anyone not using `-gsce`"? 

And any chance you've measured the impact of not always adding linkage names to subprogram DIEs for SCE lately? Is that optimization still pulling its weight for you, or would it be possible to converge on the non-sce behavior and remove some special cases here, rather than adding a new one?

**Review by OCHyams — COMMENTED** [review]
Thanks both for the reviews.

> By "upstream" you mean "anyone not using -gsce"?

Ah, yes indeed. I'll edit the summary so that it makes more sense.

> And any chance you've measured the impact of not always adding linkage names to subprogram DIEs for SCE lately? Is that optimization still pulling its weight for you, or would it be possible to converge on the non-sce behavior and remove some special cases here, rather than adding a new one?

I have - across some internal codebases and configurations I measured an increase of total linked binary size including debug info of between ~4% (-O0 builds) to as much as ~19% (thinLTO -O3 builds) when using `-dwarf-linkage-names=All` instead of `=Abstract`.

**Review by dwblaikie — APPROVED** [review]
Sounds ok, as good as how this stuff has been implemented historically at least

**Comment by pogo59** [comment]
> & not sure how this sits with @pogo59's general premise that debugger tuning shouldn't be reachable only through the tuning flags, but should be independently accessible (does independently accessible mean supported in the driver flags, or is -Xclang/cc1/mllvm flags sufficient to meet that requirement?)

What we _don't_ want is
```
if (DD->tuneForLLDB()) {
  doTheLLDBThing();
}
```
Tuning sets the defaults for a bunch of behaviors, but each of the behaviors can be individually controlled outside of the tuning parameter. That is, we could decide to abandon the tuning idea entirely, but still get exactly the same behavior for each target, if we were willing to pass the extra umpteen flags. Whether these are driver, cc1, -mllvm, or whatever, doesn't matter.

**Pull request merged**